operator/exporter.py

```diff
@@ -0,0 +1,371 @@
+            # skip comparing rest of storages in metrics[default],
+            # since storage_data_from_csi has reached its end,
+            # and it contains no more data from healthy storage-pools
+            break
+
+    metrics.provisioner.update({"pod_name": pod_name})
+    metrics.provisioner.update(json.loads(response.data)["pod"])
+    metrics.provisioner.update(pod_details)
+
+
+def set_server_data(response, metrics, pod_name, pod_details):
+    """ Updates server[storage-pool(s) & its brick(s)] related metrics"""
+
+    # Assumes CSI Pod is healthy and able to retrieve server data from its mount points,
+    # Else default values of storages will be used.
+    for storage in metrics.storages:
+        if storage["name"] in pod_name:
+
+            for brick in storage["bricks"]:
+                brick_name = brick["node"].rstrip("."+storage["name"])
+
+                if brick_name in pod_name:
+                    brick.update(json.loads(response.data)["pod"])
+                    brick.update(pod_details)
+
+
+def collect_all_metrics():
+    """ Collect all metrics data from different listening pods """
+
+    metrics = Metrics()
+
+    set_default_values(metrics)
+    set_operator_data(metrics)
+
+    pod_data = get_pod_data()
+    for pod_name, pod_details in pod_data.items():
+
+        # skip GET request to operator
+        if "operator" in pod_name:
+            metrics.operator.update({"pod_name": pod_name})
+            metrics.operator.update(pod_details)
+            continue
+
+        try:
+            response = http.request('GET', 'http://'+ pod_details["ip_address"] +':8050/_api/metrics')
+        except urllib3.exceptions.MaxRetryError:
+            logging.error(logf(
+                "Unable to reach the pod, displaying only default values",
+                pod_name=pod_name
+            ))
+
+        if response.status == 200:
+            if "nodeplugin" in pod_name:
+                set_nodeplugin_data(response, metrics, pod_name, pod_details)
+
+            if "provisioner" in pod_name:
+                set_provisioner_data(response, metrics, pod_name, pod_details)
+
+            if "server" in pod_name:
+                set_server_data(response, metrics, pod_name, pod_details)
+
+    return metrics
+
+
+def collect_and_set_prometheus_metrics():
+    """ Add all metrics data to prometheus labels """
+
+    metrics = collect_all_metrics()
+
+    # Operator Metrics
+    storage_metrics.memory_usage.labels(
+        metrics.operator["pod_name"]).set(metrics.operator["memory_usage_in_bytes"])
+    storage_metrics.cpu_usage.labels(
+        metrics.operator["pod_name"]).set(metrics.operator["cpu_usage_in_nanoseconds"])
+
+    storage_metrics.total_number_of_containers.labels(
+        metrics.operator["pod_name"]).set(metrics.operator["total_number_of_containers"])
+    storage_metrics.number_of_ready_containers.labels(
+        metrics.operator["pod_name"]).set(metrics.operator["number_of_ready_containers"])
```

Comment:
- unless all containers are ready, pod phase will not be ready
- so, need to decide whether we can drop the metrics for containers for which the pod phase is not ready?

Reply:
Containers will show `is_started: false` when not ready. In case of pod started and then stopped I thought it's better to show which containers failed out of summation. I am leaving it to @aravindavk to decide whether to display only ready info or show default values as it is now for pods and containers. Some cases like wrong `path` pods will be stuck with containerStarting status with IP being assigned. Right now we skip all those pods whose IP is not reachable or their response is not 200.